lib/model/compose.dart

```diff
@@ -0,0 +1,108 @@
+import 'dart:math';
+
+//
+// Put functions for nontrivial message-content generation in this file.
+//
+// If it's complicated enough to need tests, it should go in here.
+//
+
+// https://spec.commonmark.org/0.30/#fenced-code-blocks
+final RegExp _openingBacktickFenceRegex = (() {
+  // Allow "up to three spaces of indentation".
+  // As of Zulip Server 7.0, we don't comply with that detail:
+  //   https://chat.zulip.org/#narrow/stream/6-frontend/topic/quote-and-reply.20fence.20length/near/1588273
+  // and that's a bug.
+  const lineStart = r'^ {0,3}';
+
+  // The backticks, captured so we can see how many.
+  const backticks = r'(`{3,})';
+
+  // The "info string" plus (meaningless) leading or trailing spaces or tabs.
+  // It can't contain backticks.
+  const trailing = r'[^`]*$';
+  return RegExp(lineStart + backticks + trailing, multiLine: true);
+})();
+
+/// The shortest opening backtick fence that's longer than any in [content].
+///
+/// Expressed as a number of backticks.
+///
+/// Use this for quote-and-reply or anything else that requires wrapping
+/// Markdown in a backtick fence.
+///
+/// See the CommonMark spec, which Zulip servers should but don't always follow:
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+int getUnusedOpeningBacktickFenceLength(String content) {
+  final matches = _openingBacktickFenceRegex.allMatches(content);
+  int result = 3;
+  for (final match in matches) {
+    result = max(result, match[1]!.length + 1);
+  }
+  return result;
+}
+
+/// Wrap Markdown [content] with opening and closing backtick fences.
+///
+/// For example:
+///
+///   const str =
+///     '```javascript\n'
+///     'console.log(\'Hello world!\');\n'
+///     '```';
+///   print(wrapWithBacktickFence(content, 'quote'));
+///   // Output:
+///   //
+///   // ````quote
+///   // ```javascript
+///   // console.log('Hello world!');
+///   // ```
+///   // ````
+///
+/// This does not parse [content] to make sure its backtick fences are properly
+/// paired and nested. If they aren't, it's probably not clear what
+/// backtick-fenced output would be more reasonable anyway --
```

Comment:
Is there a way for them to be improperly paired and nested?

I'm not sure there is. I think if you take one Markdown document with some fenced blocks in it, and you shuffle around the order of the fences, you get another perfectly fine Markdown document that just means something else: some blocks may be closed implicitly, and some lines that had been opening fences with info strings are now just part of the text because they can't be closing fences.

More generally, I believe Markdown is a language where there's no such thing as malformed syntax — just syntax that doesn't mean what you intended.

Reply:
CommonMark [does not require](https://spec.commonmark.org/0.30/#fenced-code-blocks) closing fences to be paired.

> The content of the code block consists of all subsequent lines, until a closing [code fence](https://spec.commonmark.org/0.30/#code-fence) of the same type as the code block began with (backticks or tildes), and with **at least as many** backticks or tildes as the opening code fence.

> If the end of the containing block (or document) is reached and no closing code fence has been found, the code block contains all of the lines after the opening code fence until the **end of the containing block** (or document).

Our ```` ```quote ```` construct that parses its contents as Markdown is not part of CommonMark, but if it were, the only consistent nesting behavior to specify would be that a closing fence closes the outermost block it can close under these rules. Unremarkably, we don’t currently live up to that consistency (zulip/zulip#17610).